Features/task 3

JiriLojda opened the pull request.

Commit:
KA-86 Create ItemList.DatabaseLayer project

Commit:
KA-86 Move Item to new project Contracts

Commit:
KA-86 Create IItemsRepository

Commit:
KA-86 Create ItemsRepository with dummy methods

Commit:
KA-86 Make ItemsRepository methods async

Commit:
KA-86 Add dependency injection

Install Unity
Create IIoCContainer interface describing RegisterRequestScope
Create UnityAdapter as Adapter between IIoCContainer and UnityContainer
Create IoCBootstraper project with Bootstraper
Create RegisterTypes method in Bootstraper calling RegisterTypes from Bootstraper of DatabaseLayer
Configure Unity in WebApiConfig's Register method

Commit:
KA-86 Refactor IItemsRepository, ItemsRepository so methods take correct arguments

Commit:
KA-86 Refactor controler to use repository, update tests

Commit:
KA-86 Create IGuidGenerator describing GenerateGuid method

Commit:
KA-86 Create GuidGenerator in new ServiceLayer project

Commit:
KA-86 Register GuidGenerator with IGuidGenerator for IoC

Commit:
KA-86 Add IGuidGenerator dependency into constructor of controller

Update tests to inject GuidGenerator

Commit:
KA-86 Remove unnecessary usings in all files

Commit:
KA-86 Fix incorrect reference check in Put test

Commit:
KA-86 Delete AppBootstraper from file system

Commit:
KA-86 Move source and tests projects to corresponding folders

Move Source projects to Source folder and Tests projects to Tests folder
Edit csprojs and sln accordingly

Commit:
KA-86 Refactor tests to match arrange-act-assert principle

Commit:
KA-86 Rewrite UnityResolver one line method to expression bodied

Commit:
KA-86 Add empty lines between methods declarations in IItemsRepository

Commit:
KA-86 Create IBootstrapper interface

Commit:
KA-86 Implement IBootstrapper in Bootstrapers

Bootstrapers in DatabaseLayer and ServiceLayer implement IBootstrapper interface
Bootstraper in IoCBootstraper use new Bootstrapers from DatabaseLayer ServiceLayer

Commit:
KA-86 Rename IoCBootstraper.Booststraper to IoCBootstraper.Bootstrap

Commit:
KA-86 Move Unity container register and it's adapter to IoCBoostraper

Create static method to register the container and use it in app start to configure DI container
Move UnityResolver to IoCBootstraper

Commit:
KA-86 Register ItemUrlHelper in new Api.Services project

Create Bootstrapper class in Api.Services Registering ItemUrlHelper

Commit:
KA-86 Rename GuidGenerator to IdentifierService, add tests

Rename GenerateGuid to GetIdentifier

Commit:
KA-86 Create ItemStoringService, with tests, and use it in controller

Move ItemEqualityComparer to a new project Contracts.Tests.Base
Make IdentifierService internal

Commit:
KA-86 Refactor Post_ValidItem_ReturnsDummyItemAndUrl

Commit:
KA-86 Rename Bootstraper to Bootstrapper in Database and Service Layer

Commit:
KA-86 Add Async sufix to all async methods

Commit:
KA-86 Rename utils folder in ServiceLayer to Wrappers

Commit:
KA-86 Rename DatabaseLayer to Database and ServiceLayer to Services

Commit:
KA-86 Rename IoCBootstraper to DependencyInjection

Commit:
KA-86 Fix all namespaces in the solution

Commit:
KA-86 Remove unnecessary usings in the solution

Commit:
KA-86 Remove unnecessary empty lines in Item.cs

Commit:
KA-86 Use ItemStoringService in ItemController, update tests

Commit:
KA-86 Add configuration, to use CamelCasePropertyNames

Commit:
KA-86 Create IIncludeBootstraper interface in Contracts

Commit:
KA-86 Create IResolveBootrappersTypes

Commit:
KA-86 Rewrite IoC container registration

Begin IoC registration in Global.asax with BoostrapConfig.Register
BoostrapConfig calls Create(httpConfig) on statit class BootstrapFactory
in BF is initialized property Bootstrapers using BootstrappersBuilder
BB has Include method to add Boostrapers into final list of registered boostrapers
BF.Create creates WebBootstrap passing config received as argument and list of boostrapers
UWB.RegisterDependencies method calls register types on all boostrapers received in constructor
UWB.RegisterDependencies is called in BoostraperConfig

Commit:
KA-86 Rename Bootstrap folder to DependencyInjection

IIocContainer to IDependencyInjectionContainer

Commit:
KA-86 Add comment into UnityWebBootstrap

Commit:
KA-86 Rename api tests according to api methods

Commit:
KA-86 Fix BootstrappersBuilder namespace

Rename Bootstrap to DependencyInjection

Commit:
KA-86 Remove unnecessary empty lines

Commit:
KA - 86 Create new FormatterConfig and set JsonFormatter there

Commit:
KA-87 Create interface and class for getting connection strings

Commit:
KA-87 Move ConnectionStrings to Api.Services, Register it in DI

Commit:
KA-87 Add gitignore rule to exclude connectionStrings.config

Commit:
KA-87 Rename ConnectionStrings to DatabaseConfiguration

same for interface

Commit:
KA-87 Add reference to ConectionStrings.config in Web.config

Commit:
KA-87 Install MongoDB Driver nuget

Commit:
KA-87 Connect repository to MongoDB

Commit:
KA-86 Add Created and LastUpdated properties into Item

Commit:
KA-87 Update ItemEqualityComparer according to Item model changes

Commit:
KA-87 Create TimeService

Commit:
KA-87 Use TimeService in ItemStoringService

Commit:
KA-87 Update ItemStoringServiceTests to use timeService

Move TimeService to Wrappers

Commit:
KA-87 Rewrite TimeService method Now to expression body

Commit:
KA-87 Write tests for TimeService

Commit:
KA-87 Install DataAnnotations in Contracts

Commit:
KA-87 Add DataAnnotations to Item model

Commit:
KA-87 Add validation to Post and Get methods in Controler

Commit:
KA-87 Make controllers return status code when validation failed

Commit:
KA-87 Remove all unnecessary usings

PetrSvirak commented on ItemList/Source/ItemList.Api.Services/DatabaseConfiguration.cs:
```diff
@@ -0,0 +1,19 @@
+        {
+            _connectionString = ConfigurationManager.ConnectionStrings["DefaultConnection"].ConnectionString;
+        }
+        public string GetConnectionString()
```
* It should be _cached_ and it should be a _property_. It is cached, but it is not a property "wink" – `motivation:` there is no need to force a user of your contract to store the string in local variable since you cached it already.
* Also, if it is a "DefaultConnection(String)", then the property should be named `DefaultConnection[|String]` – while renaming, consider a bit more fitting name. For example `NoSql-` or `Mongo-` or simple `-`(`-ConnectionString`)

PetrSvirak commented on ItemList/Source/ItemList.Api/App_Start/WebApiConfig.cs:
```diff
@@ -10,15 +7,14 @@ public static class WebApiConfig
                 routeTemplate: "api/v1/{controller}/{id}",
                 defaults: new { id = RouteParameter.Optional }
-            );
+            );  
```
Why have committed these two changes? You have to always go through difference and be sure you commit as few changes as possible (in other each of your committed changes has to have a purpose)

PetrSvirak commented on ItemList/Source/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -27,11 +27,21 @@ public ItemsController(IItemUrlHelper itemUrlHelper, IItemsRepository repository
+        {
+            if (id == Guid.Empty)
+            {
+                return StatusCode(HttpStatusCode.NotFound);
```
* Where is a test for this?
* Why don't you use `NotFound` method? (never use `StatusCode` if there is alternative or you don't have really good reason)

PetrSvirak commented on ItemList/Source/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -27,11 +27,21 @@ public ItemsController(IItemUrlHelper itemUrlHelper, IItemsRepository repository
         {
+            if (!ModelState.IsValid)
+            {
+                return StatusCode(HttpStatusCode.BadRequest);
```
* Where are tests for this?
* Why don't you use `NotFound` method? (never use `StatusCode` if there is alternative or you don't have really good reason)

PetrSvirak commented on ItemList/Source/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -42,13 +52,21 @@ public ItemsController(IItemUrlHelper itemUrlHelper, IItemsRepository repository
         {
+            if (id == Guid.Empty)
+            {
+                return StatusCode(HttpStatusCode.NotFound);
```
* Where are tests for this?
* Why don't you use `NotFound` method? (never use `StatusCode` if there is alternative or you don't have really good reason)

PetrSvirak commented on ItemList/Source/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -42,13 +52,21 @@ public ItemsController(IItemUrlHelper itemUrlHelper, IItemsRepository repository
         {
+            if (!ModelState.IsValid)
+            {
+                return StatusCode(HttpStatusCode.BadRequest);
```
* Where are tests for this?
* Why don't you use `NotFound` method? (never use `StatusCode` if there is alternative or you don't have really good reason)

PetrSvirak commented on ItemList/Source/ItemList.Contracts/ItemList.Contracts.csproj:
```diff
@@ -35,6 +35,7 @@
       <Private>True</Private>
     </Reference>
     <Reference Include="System" />
+    <Reference Include="System.ComponentModel.DataAnnotations" />
```
This definitely does not belong here. There are usually a bit different constraints on your DB models and view models and there are usually second to none (read different per individual service) constraints on your service models. This is why transfer objects are usually `POCO`s. Also, placing validation leads to service layer (or worse, contract :wink:) leads to the impression that some user-facing messages should be encoded within the layer. It is UI's solemn responsibility to inform a user about any state of the application (and in any way) it deems appropriate.

If you wanted to decorate your model, it would have to be a class at same layer (of abstraction) as the validation is performed – a view model in `.Api` in this case. This would require you to include some sort of mapping mechanism between presentation and service layer (~ e.g. `AutoMapper`) and sort out its configuration throughout the application.

While the text above surely describes a lot of funny work, it would consume plenty of your time and that is exactly the reason why I decided to strongly advise you to use a single model (placed in `.Contracts`) within this whole task.

You should go through a bit dirtier (but not as much as it might look like) way of handling validation: Simply place it into the controller, while still using `ModelState` to transfer your error state to the user.

PetrSvirak commented on ItemList/Source/ItemList.Contracts/Models/Item.cs:
```diff
@@ -1,18 +1,26 @@
+        [StringLength(200, MinimumLength = 1)]
         public string Value { get; set; }
 
+        public DateTime Created { get; set; }
```
While it would be wrong anyway, you completely missed attribute validation on `DateTime` fields ;-)

PetrSvirak commented on ItemList/Source/ItemList.Database/Repositories/ItemsRepository.cs:
```diff
@@ -1,31 +1,38 @@
+        {
+            var client = new MongoClient(databaseConfiguration.GetConnectionString());
+            var database = client.GetDatabase("itemlist");
+            _collection = database.GetCollection<Item>("Items");
```
Make both the magical strings named (local) constants.

PetrSvirak commented on ItemList/Source/ItemList.Database/Repositories/ItemsRepository.cs:
```diff
@@ -1,31 +1,38 @@
         public async Task<Item> GetAsync(Guid id)
         {
-            return await Task.FromResult(new Item { Id = new Guid("331c43f5-11af-43a4-83d1-7d949ae5a8d7"), Value = "text3" });
+            var result = await _collection.FindAsync(item => item.Id == id);
```
Why creating a local variable instead of using expression bodied function?

PetrSvirak commented on ItemList/Source/ItemList.Services/Wrappers/TimeService.cs:
```diff
@@ -0,0 +1,11 @@
+{
+    public class TimeService : ITimeService
+    {
+
```
Empty line strikes again? :rofl:

PetrSvirak commented on ItemList/Tests/ItemList.Services.Tests/ItemStoringServiceTests.cs:
```diff
@@ -43,6 +48,7 @@ public void SetUp()
             Item itemSentToRepository = null;
             _identifierServiceMock.GetIdentifier().Returns(expectedItem.Id);
             _repositoryMock.CreateAsync(Arg.Do<Item>(item => { itemSentToRepository = item; })).Returns(Task.CompletedTask);
+            _timeService.Now().Returns(now);
```
There is a small flow in this test. No matter how manny times I call `_timeService.Now`, it will always return `now`. Add, for example, `DateTime.MinValue` as a second parameter to the `Returns` method call and fix your test.

PetrSvirak commented on ItemList/Source/ItemList.DependencyInjection/Builder/UnityWebBootstrap.cs:
```diff
@@ -0,0 +1,38 @@
+                instance.RegisterTypes(unityAdapter);
+            }
+            
+            // Register this instance, so it works like singleton
```
This is not about singleton, is it? It is more about all contracts being registered within the container.

It won't be too nice test, however, give up to 3/4 of an hour to creating simple integration test that will gather all bootstrappers (this can be easily achieved by marking the static enumerable collection in the `Bootstrap` class `internal` (with appropriate comment on the purpose, of course)), then iterate through them cumulating registered types (this is a bit "nasty" part requiring non-trivial mock of the `IDependencyInjectionContainer`) and then check that all interfaces from `Contracts` have an implementation (this is a "nasty" part requiring some reflection). (Then fix this comment :))

PetrSvirak commented on ItemList/Source/ItemList.Database/Repositories/ItemsRepository.cs:
```diff
@@ -1,31 +1,38 @@
+
+        public ItemsRepository(IDatabaseConfiguration databaseConfiguration)
+        {
+            var client = new MongoClient(databaseConfiguration.GetConnectionString());
```
Read more about `MongoClient` instance management in the official document. You use it as a request-scoped dependency which is not entirely correct.

PetrSvirak reviewed (changes requested):
You seem to work in quite a hurry and this seems to have led, quite unfortunately, to some serious oversights from my point of view. Also, this task requires you to only have `Post` and `Get` methods fully working - `Controller` without other than these 3 actions (or with them being dummy or unsanitized) is sufficient :wink:

I see that time you put into design finally paid off tough as you finished the task with quite a few changes.

There are some issues bellow that need to be addressed,
most severe seems to be missing tests - if you happen not to see the purpose of the (either missing or present) tests clearly, just ask, we can discuss the topic further.

BTW: 5c657fe has improper task number :wink:

Commit:
KA-87 Add property DefaultConnectionString instead of old method

Commit:
KA-87 Remove data anotattions from Item

Commit:
KA-87 Add custom validation into ItemsController, update tests

Commit:
KA-87 Remove unnecessary string parameter from DatabaseConfiguration

Unnecessary string parameter in DatabaseConfiguration constructor was breaking DI

Commit:
KA-87 Make database, collection names local consts in ItemsRepository

Commit:
KA-87 Refactor GetAllAsync to expression body function

Commit:
KA-87 Remove empty line in TimeService

Commit:
KA-87 Use same DateTime for Item created&updated properties,test it

Commit:
KA-87 Add,implement and use RegisterSingleton for Database Bootsraper

Commit:
KA-87 Create UnityWebResolverBuilder, registrating types

UnityWebResolverBuilder calls given methods with UnityAdapter
UnityWebResolverBuilder sets DependencyResolver property on HttpConfiguration given in ctor
Bootstrap gives RegisterTypes from all Bootstrappers to resolverBuilder.RegisterDependencies and then calls RegisterDependencyResolver

Commit:
KA-87 Polish AllContractsHaveRegisteredImplementation

Move ContainerMock into separate file

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContainerMock.cs:
```diff
@@ -0,0 +1,28 @@
+            where TImplementation : TType
+            => AddTypeFullName<TType>();
+
+        private void AddTypeFullName<TType>() => _registeredContracts.Add(typeof(TType).FullName);
```
Private methods should be at the end of file

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContainerMock.cs:
```diff
@@ -0,0 +1,28 @@
+{
+    internal class ContainerMock : IDependencyInjectionContainer
+    {
+        private readonly ICollection<string> _registeredContracts = new List<string>();
```
Consider using `Set` instead of `List`. Or even better, make the mock "real wrapper" and let it accept an `ICollection` as its dependency in the constructor (no `null-check` needed as the class is intended for tests).

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -0,0 +1,73 @@
+    [TestFixture]
+    internal class ContratsImplementationTests
+    {
+        private ContainerMock _containerMock;
```
How cool would it look like if it were `IDependencyInjectionContainer` instead of `ContainerMock`

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -0,0 +1,73 @@
+        public void SetUp()
+        {
+            _containerMock = new ContainerMock();
+            _resolverMock = Substitute.For<IResolverBuilder>();
```
Consider moving all `IResolverBuilder` mocking into a standalone method.

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -0,0 +1,73 @@
+        public void AllContractsHaveRegisteredImplementation()
+        {
+            var bootstrap = new Bootstrap(BootstrapFactory.Bootstrappers.Value, _resolverMock);
+            Assembly contractsAssembly = typeof(IDependencyInjectionContainer).Assembly;
```
Move arrangement of expected contracts (mind the name I used here :wink:) to a standalone method so the test is clearer.

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -0,0 +1,73 @@
+                .Where(type => type.IsInterface)
+                .Select(type => type.FullName)
+                .ToList();
+            // container should not register itself
```
Place an empty line between a comment and code above (that does not belong to the comment) – it will say more clearly what does it refer to.

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -0,0 +1,73 @@
+            var actualRegisteredContracts = _containerMock.RegisteredContracts;
+
+            const string firstSeparator = ",\n\t\t\t\t\t\t\t\t   ";
+            const string secondSeparator = ",\n\t\t\t\t\t\t  ";
```
Whaaat: :rofl:

PetrSvirak replied:
It's quite a separator. Use a combination of `String.Concat` `Enumerable.Repeat` or even better [verbatim string](http://stackoverflow.com/questions/3311988/what-is-the-difference-between-a-regular-string-and-a-verbatim-string).

Did you know section: Did you know that is is possible to combine verbatim and interpolated strings? :)

Commit:
KA-87 Repair NuGet packages location in Api.csproj

PetrSvirak commented on ItemList/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -0,0 +1,73 @@
+            Assert.That(
+                actualRegisteredContracts,
+                Is.EquivalentTo(expectedRegisteredContracts),
+                $"This should not be registered: [ {string.Join(firstSeparator, actualRegisteredContracts.Except(expectedRegisteredContracts))} ],\n\n" +
```
Please, move the code generating the message to a separate method.

PetrSvirak commented on ItemList/ItemList.sln:
```diff
@@ -25,6 +25,8 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ItemList.Services.Tests", "
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Contracts.Tests.Base", "Tests\Contracts.Tests.Base\Contracts.Tests.Base.csproj", "{B016245C-CC90-45D5-9343-74A4B3B28731}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "ItemList.DependencyInjection.Tests", "ItemList.DependencyInjection.Tests\ItemList.DependencyInjection.Tests.csproj", "{D93A9F9A-F6A1-4E8B-8EF6-53CA07EB871F}"
```
Move the project to the `Tests` folder :wink:

PetrSvirak commented on ItemList/Source/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -15,6 +16,26 @@ public class ItemsController : ApiController
         private readonly IItemsRepository _repository;
         private readonly IItemStoringService _itemStoringService;
 
+        private void ValidateItem(Item item, ModelStateDictionary modelState)
```
https://imgflip.com/i/1ng0pn

PetrSvirak commented on ItemList/Source/ItemList.Database/Repositories/ItemsRepository.cs:
```diff
@@ -11,18 +11,17 @@ namespace ItemList.Database.Repositories
     {
         private readonly IMongoCollection<Item> _collection;
+        private const string CollectionName = "Items";
+        private const string DatabaseName = "itemlist";
```
Did you know section: Did you know you can actually let `MongoDB` extract database name from the connection string? See [MongoUrl](http://api.mongodb.com/csharp/current/html/T_MongoDB_Driver_MongoUrl.htm) for more information.

PetrSvirak commented on ItemList/Source/ItemList.DependencyInjection/Adapters/UnityAdapter.cs:
```diff
@@ -22,5 +22,15 @@ public void RegisterRequestScoped<TType>(Func<TType> implementationFactory)
+
+        public void RegisterSingleton<TInterface, TImplementation>() where TImplementation : TInterface
+        {
+            _container.RegisterType<TInterface, TImplementation>(new ContainerControlledLifetimeManager());
```
You actually repeat yourself here – consider creating a helper (~ private) method that would accept 2 generic arguments and a lifetime manager…

PetrSvirak commented on ItemList/Source/ItemList.DependencyInjection/Adapters/UnityAdapter.cs:
```diff
@@ -22,5 +22,15 @@ public void RegisterRequestScoped<TType>(Func<TType> implementationFactory)
+
+        public void RegisterSingleton<TType>(Func<TType> implementationFactory)
+        {
+            _container.RegisterInstance(new InjectionFactory(_ => implementationFactory()));
```
Huh. `RegisterInstance` actually expects an instance of the served object. Not that this would not work, apparently, however, it is not intended use of the method. To use `RegisterInstance` properly, you should have called `implementationFactory()` and put its result into the `RegisterInstance` method.

When using `InjectionFactory`, you should use `RegisterType` – in this case, same as for the other overload of `RegisterSingleton` interface will apply.

PetrSvirak reviewed (changes requested):
It's almost done. You made some outstanding design during the whole task – that is really great, thank you for that. 

Unfortunately, there are still some things missing though (especially the first two bullets are somewhat more essential for an API, sometimes "one does not see the forest for all the trees"):
* Your `Get` action returns OK 200 even if an item does not exists – write a test for it and fix this behavior
* Your `Post` action receives an instance of a reference type (~ class `Item`). Such type might become `null` – imagine the case somebody sends to your `Post` complete nonsense (e.g. JSON for another API). Write a test for such situation and fix your action's behavior.
* Your `Delete` and `Put` actions deserve a short comment saying they are still considered stub
* There are also some inline comments worthy of your attention.

Commit:
KA-87 Return NotFound on Get with nonexisting Id, check null in Post

GetAsync returns NotFound when given nonexisting Id
PostAsync checks for item being null or having null values
Update tests accordingly

Commit:
KA-87 Add comments to dummy method in controller

Commit:
KA-87 Move private ValidateItem method under public methods

Commit:
KA-87 Make ContainerMock accept registeredContract collection in ctor

Move private method AddTypeFullName under public methods

Commit:
KA-87 Extract setup of resolverMock into SetUpResolverMock method

Commit:
KA-87 Extract private method GetExpectedContracts

Add empty line between comment and code above that does not belong to the comment
Make empty lines strike again :]

Commit:
KA-87 Create CreateSeparator private method, Use verbatim strings

CreateSeparator concates specified number of tabs and spaces to given initialString

Commit:
KA-87 Create GetErrorMessage private static method

GetErrorMessage returns error message for given actual and expected contracts

Commit:
KA-87 Move DependencyInjection.Tests project to the Tests folder

Commit:
KA-87 Polish GetErrorMessage method

Commit:
KA-87 Use MongoUrl in ItemsRepository to get DatabaseName

Commit:
KA-87 Use RegisterType instead of RegisterInstance in UnityAdapter

Commit:
KA-87 Rewrite IDependencyInjection as fluent api

Update UnityAdapter and ContainerMock accordingly
Create helper method in UnityAdapter to solve code repetition

PetrSvirak commented on ItemList/Source/ItemList.DependencyInjection/Adapters/UnityAdapter.cs:
```diff
@@ -12,25 +12,31 @@ public UnityAdapter(UnityContainer container)
 
-        public void RegisterSingleton<TInterface, TImplementation>() where TImplementation : TInterface
+        public IDependencyInjectionContainer RegisterSingleton<TType>(Func<TType> implementationFactory) 
+            => RegisterType<TType>(new ContainerControlledLifetimeManager(), new InjectionFactory(_ => implementationFactory()));
```
You actually call `new InjectionFactory(_ => implementationFactory()` in both `RegisterSingleton` and `RegisterRequestScoped` methods, don't you? This goes against the purpose of the `RegisterType` helper method – please, move this piece of code down to the helper method.

PetrSvirak commented on ItemList/Source/ItemList.Api.Services/Bootstrapper.cs:
```diff
@@ -10,9 +10,10 @@ public class Bootstrapper : IBootstrapper
-            container.RegisterRequestScoped<IItemUrlHelper, ItemUrlHelper>();
-            container.RegisterRequestScoped(() => (HttpRequestMessage)HttpContext.Current.Items["MS_HttpRequestMessage"]);
-            container.RegisterRequestScoped<IDatabaseConfiguration, DatabaseConfiguration>();
+            container
```
This is really great. You can make these [expression bodied function](https://imgflip.com/i/1nlubh), can't you?

PetrSvirak commented on ItemList/Source/ItemList.Api/Controllers/ItemsController.cs:
```diff
@@ -71,17 +54,43 @@ public ItemsController(IItemUrlHelper itemUrlHelper, IItemsRepository repository
+            }
+            if (item.Value != null && item.Value.Length > 200)
+            {
+                modelState.AddModelError(nameof(item.Value), "Value cannot exceed 200 characters.");
```
Just a detail, but you always use "should (not)"; consistency is nice.

PetrSvirak commented on ItemList/Source/ItemList.Database/Repositories/ItemsRepository.cs:
```diff
@@ -12,12 +12,12 @@ public class ItemsRepository : IItemsRepository
 
         public ItemsRepository(IDatabaseConfiguration databaseConfiguration)
         {
+            var databaseName = MongoUrl.Create(databaseConfiguration.DefaultConnectionString).DatabaseName;
```
You can actually store the `mongoUrl` and use it the `MongoClient` constructor (and then use its `DatabaseName` property).

PetrSvirak commented on ItemList/Tests/ItemList.DependencyInjection.Tests/ContainerMock.cs:
```diff
@@ -0,0 +1,37 @@
+            where TImplementation : TType
+            => AddTypeFullName<TType>();
+        
+        public IDependencyInjectionContainer RegisterSingleton<TType>(Func<TType> implementationFactory) => AddTypeFullName<TType>();
```
New line before `=>` :wink:

PetrSvirak commented on ItemList/Tests/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -19,7 +19,65 @@ internal class ContratsImplementationTests
+        }
+
+        [Test]
+        public void AllContractsHaveRegisteredImplementation()
```
The naming of tests should be consistent. In this case, something along the lines of `Bootstrap_KnownBootstrapper_RegistersAllContracts` would be a bit more fitting.

PetrSvirak commented on ItemList/Tests/ItemList.DependencyInjection.Tests/ContratsImplementationTests.cs:
```diff
@@ -19,7 +19,65 @@ internal class ContratsImplementationTests
+                );
+        }
+
+        private static string GetErrorMessage(ICollection<string> actualContracts, ICollection<string> expectedContracts)
```
You spent quite some time on formatting the message. I appreciate your effort, yet these are tests for you (and other developers of course) and most of us do not see well through pink UX glasses. The code is quite complex, from my point of view and does not cope well with different sizes of tests results window, however, if you see benefit in it (or at least you feel you really put quite some effort to it as I do feel), do not change it (since it is your code, under your production :wink:)

PetrSvirak reviewed (changes requested):
OK, it is all quite great, now. Really. There are, however, 5 little details I would normally overlook and 1 little detail that cannot be overlooked so easily. So, let's fix it quickly and quietly:man_cartwheeling:

Commit:
KA-87 Move InjectionFactory creation to helper method

Commit:
KA-87 Convert RegisterTypes into lambda function

Commit:
KA-87 Rewrite item validation more consistently

Commit:
KA-87 Use another MongoClient ctor to shorten code

Commit:
KA-87 Polish ContainerMock: spacing

Commit:
KA-87 Rename test to be more consistent

AllContractsHaveRegisteredImplementation to Bootstrap_KnownBootstrapper_RegistersAllContracts

PetrSvirak committed:
KA-86 Unify error messages in ItemsController

PetrSvirak commented:
All good, I just made a small approval change regards the error messages in Item's controller, otherwise, it is now all great. Let's merge it.
